mm2src/coins/qrc20/history.rs

```diff
@@ -0,0 +1,991 @@
+use super::*;
+use crate::utxo::{RequestTxHistoryResult, UtxoFeeDetails};
+use crate::CoinsContext;
+use crate::TxFeeDetails;
+use byteorder::{BigEndian, ReadBytesExt, WriteBytesExt};
+use common::jsonrpc_client::JsonRpcErrorType;
+use common::lazy::LazyLocal;
+use common::mm_metrics::MetricsArc;
+use futures01::Future as Future01;
+use itertools::Itertools;
+use script_pubkey::{extract_contract_call_from_script, extract_gas_from_script, ExtractGasEnum};
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::future::Future;
+use std::io::Cursor;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::thread;
+use std::time::Duration;
+use utxo_common::{HISTORY_TOO_LARGE_ERROR, HISTORY_TOO_LARGE_ERR_CODE};
+
+type TxTransferMap = HashMap<TxInternalId, TransactionDetails>;
+type HistoryMapByHash = HashMap<H256Json, TxTransferMap>;
+type TxIds = Vec<(H256Json, u64)>;
+
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct TxInternalId {
+    tx_hash: H256Json,
+    output_index: i64,
+    log_index: i64,
```

Comment:
Is there a reason to use i64? I'm a bit confused as I think that output_index and log_index can't ever be negative :slightly_smiling_face:

Reply:
I tried to get history and tx receipt for just sent transaction. The results do not contain info about not mined transaction, because of this an electrum should return only positive values. I switch the signed fields with unsigned, even height, transaction_index etc.